src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Dequeue()
+        {
+            if (this.TryDequeue(out var element, out var priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    "The priority queue is empty, cannot dequeue an element.");
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryDequeue([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (this.nodes.Count == 0)
+            {
+                element = default;
+                priority = default;
+                return false;
+            }
+            else
+            {
+                (element, priority) = this.nodes[RootIndex];
+                this.Remove(RootIndex);
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryPeek([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (this.nodes.Count == 0)
+            {
+                element = default;
+                priority = default;
+                return false;
+            }
+            else
+            {
+                (element, priority) = this.nodes[RootIndex];
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Combined enqueue/dequeue operation, generally more efficient than sequential Enqueue/Dequeue calls.
+        /// </summary>
+        public TElement EnqueueDequeue(TElement element, TPriority priority)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Enqueues a collection of element/priority pairs.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<(TElement Element, TPriority Priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            foreach (var (element, priority) in items)
+            {
+                this.Enqueue(element, priority);
+            }
```

Comment:
Presumably we can do better than individually queueing each element, no? i.e. add them all and then heapify once?  Otherwise, I don't think we should have the API.

Reply:
If we have a heap with `n` elements and we want to `EnqueueRange` `m` elements, then the complexity of invoking heapify would be `O(n+m)`. This could make sense if `m` is much larger than `n`, but this sounds like something that has very limited usefulness, could be considered if the community sees the need.

Approved to remove this API, and the similar method below where you commented *Ditto*?

Reply:
> then the complexity of invoking heapify would be O(n+m)

So wouldn't it be better to have this be `O(n + m)` rather than the `O(n * m)` you get by calling `Enqueue` `m` times?  Other than writing a few more lines of code, what's the downside... is that significantly more expensive if `m` is small?  And if so, doesn't that also mean the ctor that takes an enumerable is problematic and that it should instead just call Enqueue `m` times?  The inconsistency here is part of what bothers me.

Reply:
Hmm, I think it's like this:

* If `m` is large compared to `n`, then the approach with `Heapify` would be better and then `EnqueueRange` would have complexity `O(n + m)`.
* If `m` is small compared to `n`, then the approach with calling `Enqueue` `m` times would be better and then `EnqueueRange` would have complexity `O(m * log n)`.

Each execution of `Enqueue` costs us `O(log n)`.

What do you suggest then? Perhaps the choice of the strategy should be dynamic, but then we would need to conduct a study of how to determine the threshold. Perhaps it's out of scope for this PR (unless someone is willing to jump in and check it) and we can for now either fix this to always use one of the strategies.

Reply:
> What do you suggest then?

The simplest optimization would be to use heapify if the current `_size_` is zero. For any other scenaria we would probably need to check that the input IEnumerable is ICollection (and thus obtain the count before enumerating), then decide whether heapify is worth it.

Reply:
Would you be then willing to conduct a study of when heapify is worth it?